### PR TITLE
[opt](memory) Add GlobalMemoryArbitrator and support ReserveMemory

### PR DESCRIPTION
**Pull request by xinyiZzz**
## Proposed changes

Add GlobalMemoryArbitrator, used for check memory exceed limit, ReserveMemory

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/runtime/memory/global_memory_arbitrator.h [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+#include "util/mem_info.h"
+
+namespace doris {
+
+class GlobalMemoryArbitrator {
+public:
+    static inline int64_t proc_mem_no_allocator_cache() {
+        return _s_proc_mem_no_allocator_cache.load(std::memory_order_relaxed);
+    }
+    static inline int64_t proc_mem_corrected() {
```
Comment (yiguolei) [inline be/src/runtime/memory/global_memory_arbitrator.h:29]: add comment to explain the function。

File: be/src/runtime/memory/thread_mem_tracker_mgr.h [hunk 2]
```diff
@@ -164,16 +172,35 @@ inline bool ThreadMemTrackerMgr::push_consumer_tracker(MemTracker* tracker) {
...
 inline void ThreadMemTrackerMgr::pop_consumer_tracker() {
     DCHECK(!_consumer_tracker_stack.empty());
     _consumer_tracker_stack.back()->consume(_untracked_mem);
+    _consumer_tracker_stack.back()->release(_reserved_mem);
     _consumer_tracker_stack.pop_back();
 }
 
 inline void ThreadMemTrackerMgr::consume(int64_t size, int skip_large_memory_check) {
+    if (_reserved_mem != 0) {
+        if (_reserved_mem >= size) {
```
Comment (yiguolei) [inline be/src/runtime/memory/thread_mem_tracker_mgr.h:188]: 这里要加注释

File: be/src/runtime/memory/thread_mem_tracker_mgr.h [hunk 3]
```diff
@@ -209,14 +236,58 @@ inline void ThreadMemTrackerMgr::flush_untracked_mem() {
...
+    _limiter_tracker_raw->consume(_old_untracked_mem);
+    for (auto* tracker : _consumer_tracker_stack) {
+        tracker->consume(_old_untracked_mem);
     }
-    _untracked_mem -= old_untracked_mem;
+    _untracked_mem -= _old_untracked_mem;
     _stop_consume = false;
 }
 
+inline bool ThreadMemTrackerMgr::try_reserve(int64_t size, bool force_tracker_overcommit) {
```
Comment (yiguolei) [inline be/src/runtime/memory/thread_mem_tracker_mgr.h:251]: why need overcommit?

File: be/src/common/daemon.cpp [hunk 4]
```diff
@@ -247,33 +248,35 @@ void Daemon::memory_gc_thread() {
...
         auto sys_mem_available = doris::MemInfo::sys_mem_available();
-        auto proc_mem_no_allocator_cache = doris::MemInfo::proc_mem_no_allocator_cache();
+        auto proc_mem_corrected = doris::GlobalMemoryArbitrator::proc_mem_corrected();
 
         // GC excess memory for resource groups that not enable overcommit
-        auto tg_free_mem = doris::MemInfo::tg_not_enable_overcommit_group_gc();
+        auto tg_free_mem = doris::MemInfo::tg_disable_overcommit_group_gc();
         sys_mem_available += tg_free_mem;
-        proc_mem_no_allocator_cache -= tg_free_mem;
+        proc_mem_corrected -= tg_free_mem;
```
Comment (yiguolei) [inline be/src/common/daemon.cpp:256]: proc_mem_corrected  这个名字太难懂了

Comment (xinyiZzz) [reply be/src/common/daemon.cpp:256]: modify to `process_memory_usage`

File: be/src/runtime/memory/global_memory_arbitrator.h (hunk 1 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/global_memory_arbitrator.h:29]: done

File: be/src/runtime/memory/thread_mem_tracker_mgr.h (hunk 2 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/thread_mem_tracker_mgr.h:188]: done

File: be/src/runtime/memory/thread_mem_tracker_mgr.h (hunk 3 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/thread_mem_tracker_mgr.h:251]: default query/load allows overcommit, if `force_tracker_overcommit` is false, try_consume will not check the mem_limit of MemTracker.

Comment (xinyiZzz) [reply be/src/runtime/memory/thread_mem_tracker_mgr.h]: delete `force_tracker_overcommit`

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

File: be/src/runtime/memory/thread_mem_tracker_mgr.h [hunk 5]
```diff
@@ -209,14 +243,58 @@ inline void ThreadMemTrackerMgr::flush_untracked_mem() {
...
+    _untracked_mem -= _old_untracked_mem;
     _stop_consume = false;
 }
 
+inline bool ThreadMemTrackerMgr::try_reserve(int64_t size) {
+    DCHECK(_limiter_tracker_raw);
+    DCHECK(size >= 0);
+    CHECK(init());
+    flush_untracked_mem();
+    DCHECK(_reserved_mem == 0);
```
Comment (yiguolei) [inline be/src/runtime/memory/thread_mem_tracker_mgr.h:263]: 不应该在这dcheck，一个线程里reserve 两次是合理的

File: be/src/runtime/memory/thread_mem_tracker_mgr.h [hunk 6]
```diff
@@ -209,14 +243,58 @@ inline void ThreadMemTrackerMgr::flush_untracked_mem() {
...
+    _limiter_tracker_raw->consume(_old_untracked_mem);
+    for (auto* tracker : _consumer_tracker_stack) {
+        tracker->consume(_old_untracked_mem);
     }
-    _untracked_mem -= old_untracked_mem;
+    _untracked_mem -= _old_untracked_mem;
     _stop_consume = false;
 }
 
+inline bool ThreadMemTrackerMgr::try_reserve(int64_t size) {
```
Comment (yiguolei) [inline be/src/runtime/memory/thread_mem_tracker_mgr.h:258]: 应该是在切换memtracker的时候，如果发现上一个memtracker的reserved bytes != 0，那么应该直接DCHECK 挂掉。

File: be/src/runtime/memory/thread_mem_tracker_mgr.h (hunk 5 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/thread_mem_tracker_mgr.h:263]: done

File: be/src/runtime/memory/thread_mem_tracker_mgr.h (hunk 6 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/thread_mem_tracker_mgr.h:258]: image
加了个 DCHECK，不过不可能发生你说的情况，因为 reserve 必须在 scoped attach task 内，scoped结束时 detach task 时会调 `release_reserved()`

**Comment by xinyiZzz** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by mrhhsg — APPROVED** [review]
LGTM

**Pull request merged**